codegen/masm/src/codegen/scheduler.rs

```diff
@@ -0,0 +1,1404 @@
+    // beyond the current block, as those are dead code and should be eliminated in the DCE step.
+    //
+    // The actual scheduling decision for the instruction is deferred to `analyze_inst`, where we
+    // treat the instruction similarly to argument materialization, and either make it a pre-requisite
+    // of the instruction or execute it in the post-execution phase depending on the terminator type
+    let terminator = {
+        let block = function.dfg.block(block_id);
+        let id = block.last().unwrap();
+        Node::Inst {
+            id,
+            pos: block.len() as u16,
+        }
+    };
+    let terminator_id = terminator.into();
+    for (inst_index, inst) in function.dfg.block_insts(block_id).enumerate() {
+        let opcode = function.dfg.inst(inst).opcode();
+        // Skip the block terminator
+        if opcode.is_terminator() {
+            continue;
+        }
+
+        let node = Node::Inst {
+            id: inst,
+            pos: inst_index as u16 + 1,
+        };
+        let node_id = node.id();
+
+        // Skip instructions with transitive dependents on at least one result, or a direct dependent
+        let has_dependents = graph.predecessors(node_id).any(|pred| {
+            if pred.dependent.is_result() {
+                graph.num_predecessors(pred.dependent) > 0
+            } else {
+                true
+            }
+        });
+        if has_dependents {
+            continue;
+        }
+
+        // Instructions with no side effects require a control dependency if at least
+        // one result is live after the end of the current block. We add the dependency
+        // to the instruction results if present, otherwise to the instruction itself.
+        let pp = ProgramPoint::Block(block_id);
+        let mut live_results = SmallVec::<[NodeId; 2]>::default();
+        for pred in graph.predecessors(node_id) {
+            match pred.dependent.into() {
+                Node::Result { value, .. } => {
+                    let is_live_after = liveness.is_live_after(&value, pp);
+                    if is_live_after {
+                        live_results.push(pred.dependent);
+                    }
+                }
+                _ => continue,
+            }
+        }
+
+        let has_live_results = !live_results.is_empty();
+        for result_node in live_results.into_iter() {
+            graph.add_dependency(terminator_id, result_node);
+        }
+
+        // Instructions with side effects but no live results require a control dependency
+        if opcode.has_side_effects() && !has_live_results {
+            graph.add_dependency(terminator_id, node_id);
+            continue;
+        }
+    }
+
+    dce(graph, block_id, function, liveness)
+}
+
+fn add_data_dependency(
+    dependent_id: NodeId,
+    dependency_id: NodeId,
+    value: hir::Value,
+    pp: ProgramPoint,
+    function: &hir::Function,
+    graph: &mut DependencyGraph,
+) {
```

Comment:
I'm not sure how generic this function really is (as in, how useful it would be to another consumer of `DependencyGraph`), but you have a point that it probably could live alongside `DependencyGraph` rather than here. I'll move it, we can always revisit in the future if it turns out that it is too specific to be used elsewhere.